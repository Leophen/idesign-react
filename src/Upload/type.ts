export interface UploadProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义上传内容
   */
  children?: React.ReactNode;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 占位符
   */
  placeholder?: string;
  /**
   * 上传组件风格
   * @default button
   */
  theme?: 'button' | 'block' | 'drag';
  /**
   * 文件大小限制，单位 M
   * @default 10
   */
  maxSize?: number;
  /**
   * 上传时触发
   */
  onChange?: (file: File) => void;
}
