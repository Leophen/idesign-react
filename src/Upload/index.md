---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 12
---

# Upload 上传

用户上传任意内容的组件。

## 基本用法

```tsx
import React from 'react';
import { Upload } from 'idesign-react';

const App = () => {
  const handleChange = (file) => {
    console.log(file)
  }

  return (
    <Upload onChange={handleChange} />
  );
};

export default App;
```

## 不同主题

可通过 `theme` 属性控制不同属性的上传组件：

```tsx
import React from 'react';
import { Upload } from 'idesign-react';

const App = () => {
  const handleChange = (file) => {
    console.log(file)
  }

  return (
    <div className="idesign-demo-block-row">
      <Upload theme="button" onChange={handleChange} />
      <Upload theme="block" onChange={handleChange} />
      <Upload theme="drag" onChange={handleChange} />
    </div>
  );
};

export default App;
```

## 自定义提示

可通过 `placeholder` 属性控制自定义占位符：

```tsx
import React from 'react';
import { Upload } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <Upload placeholder="自定义占位符" />
      <Upload placeholder="自定义占位符" />
      <Upload placeholder="自定义占位符" />
    </div>
  );
};

export default App;
```

## 自定义节点

可通过包裹自定义节点来展示任意上传组件：

```tsx
import React from 'react';
import { Upload } from 'idesign-react';

const App = () => {
  return (
    <Upload>
      <img width={100} src="/images/logo.svg" />
    </Upload>
  );
};

export default App;
```

<API />
