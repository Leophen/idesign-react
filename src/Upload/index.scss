

.i-upload {
  color: var(--i-font-2);
  font-size: 14px;
}

.i-upload__block {
  display: flex;
  flex-direction: column;
  gap: 8px;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 100px;
  height: 100px;
  border: 1px dashed var(--i-border);
  border-radius: 8px;
  cursor: pointer;
  transition: all 0.2s ease-out;
  &:hover {
    border-color: var(--i-border-hover);
  }
}

.i-upload__drag {
  display: flex;
  flex-direction: column;
  gap: 6px;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 320px;
  padding: 12px 0;
  background: var(--i-bg-hover);
  border: 1px dashed var(--i-border);
  border-radius: 8px;
  cursor: pointer;
  transition: all 0.2s ease-out;
  &:hover {
    border-color: var(--i-border-hover);
  }
  &.i-upload__dragging {
    border-color: var(--i-primary);
  }
}

.i-upload-placeholder {
  color: var(--i-font-4);
}

.i-upload-tip {
  color: var(--i-font-5);
  font-size: 12px;
}
