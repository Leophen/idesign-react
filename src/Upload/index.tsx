import React, { useMemo, useRef, useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import Button from '../Button';
import Icon from '../Icon';
import Message from '../Message';
import { UploadProps } from './type';

const Upload: React.FC<UploadProps> = (props) => {
  const defaultPlaceHolder = props.theme === 'drag' ? '点击或拖动到框内上传' : '点击上传'
  const {
    children,
    className,
    style,
    placeholder = defaultPlaceHolder,
    theme = 'button',
    maxSize = 10,
    onChange,
    ...restProps
  } = props;

  const MAX_SIZE = useMemo(() => maxSize * 1024 * 1024, [maxSize])

  const inputRef = useRef<HTMLInputElement>(null)

  const handleClickInput = () => {
    inputRef.current?.click()
  }

  const handleUpload = (e: React.ChangeEvent) => {
    e.persist()
    const file = ((e.target as HTMLInputElement).files || [])[0]
    if (file) {
      if (file.size > MAX_SIZE) {
        const unit = MAX_SIZE > 1 * 1024 ? 'M' : 'kb'
        const size = MAX_SIZE > 1 * 1024 ? maxSize : maxSize * 1024
        Message.error(`文件大小不得超过 ${size} ${unit}`);
      } else {
        onChange?.(file)
      }
    }
  }

  const [ifDragIn, setIfDragIn] = useState(false)

  const handleDragOver = (e: React.MouseEvent) => {
    e.preventDefault()
    setIfDragIn(true)
  }

  const handleDragLeave = (e: React.MouseEvent) => {
    setIfDragIn(false)
  }

  const handleDragDrop = (e: React.MouseEvent) => {
    e.preventDefault()
    handleClickInput()
    setIfDragIn(false)
  }

  const renderUpload = () => {
    if (theme === 'button') {
      return (
        <Button type="info" variant="outline" icon="ArrowLineUpload">
          {placeholder}
        </Button>
      )
    } else if (theme === 'block') {
      return (
        <div className="i-upload__block">
          <Icon name="ThePlus" size={20} />
          {placeholder}
        </div>
      )
    } else if (theme === 'drag') {
      return (
        <div
          className={classNames(
            'i-upload__drag',
            ifDragIn && 'i-upload__dragging'
          )}
          draggable
          onDragOver={handleDragOver}
          onDragLeave={handleDragLeave}
          onDrop={handleDragDrop}
        >
          <Icon name="ArrowLineUpload" size={20} />
          <span className='i-upload-placeholder'>{placeholder}</span>
          <span className='i-upload-tip'>支持不超过 10M 的 excel 类型文件</span>
        </div>
      )
    }
  }

  return (
    <div
      className={classNames(
        'i-upload',
        className
      )}
      style={{ ...style }}
      onClick={handleClickInput}
      {...restProps}
    >
      <input
        type="file"
        style={{ display: 'none' }}
        ref={inputRef}
        onChange={handleUpload}
      />
      {children ? children : renderUpload()}
    </div>
  );
};

Upload.displayName = 'Upload';

export default Upload;
