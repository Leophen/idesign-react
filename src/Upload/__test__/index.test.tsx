import { render } from '@testing-library/react';
import Upload from '../index';

describe('Upload 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Upload />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('block', () => {
    const { asFragment } = render(<Upload theme="block" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('drag', () => {
    const { asFragment } = render(<Upload theme="drag" />);
    expect(asFragment()).toMatchSnapshot();
  });
});

