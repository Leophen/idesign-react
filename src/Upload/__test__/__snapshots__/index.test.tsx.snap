// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload 组件测试 block 1`] = `
<DocumentFragment>
  <div
    class="i-upload"
  >
    <input
      style="display: none;"
      type="file"
    />
    <div
      class="i-upload__block"
    >
      <i
        class="i-icon icon-ThePlus"
        style="font-size: 20px;"
      />
      点击上传
    </div>
  </div>
</DocumentFragment>
`;

exports[`Upload 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-upload"
  >
    <input
      style="display: none;"
      type="file"
    />
    <button
      class="i-button i-button--type-info i-button--variant-outline"
    >
      <i
        class="i-icon icon-ArrowLineUpload i-button-icon"
      />
      点击上传
    </button>
  </div>
</DocumentFragment>
`;

exports[`Upload 组件测试 drag 1`] = `
<DocumentFragment>
  <div
    class="i-upload"
  >
    <input
      style="display: none;"
      type="file"
    />
    <div
      class="i-upload__drag"
      draggable="true"
    >
      <i
        class="i-icon icon-ArrowLineUpload"
        style="font-size: 20px;"
      />
      <span
        class="i-upload-placeholder"
      >
        点击或拖动到框内上传
      </span>
      <span
        class="i-upload-tip"
      >
        支持不超过 10M 的 excel 类型文件
      </span>
    </div>
  </div>
</DocumentFragment>
`;
